ユーザーインクリメンタルサーチ

Hiro3nori3 opened the pull request:
# WHAT 
ユーザー検索をインクリメンタルサーチ化する

Commit:
routes.rbの編集

・ルーティングにsearch機能のための記述を追加

Commit:
users_controllerの作成とsearchファンクションの追加

・users_controllerを作成した
・searchアクションを定義した

Commit:
search.json.jbuilderの作成

・search.json.jbuilderを作成した

Commit:
groups/new.html.hamlの編集

・インクリメンタルサーチ機能のための記述をビューに用意した

Commit:
search.jsの作成

・ajaxを用いた非同期通信で、ビューにユーザーの一覧を表示できるようにした

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,41 @@
+$(function() {
+	var user_list = $("#user-search-result")
```
変数名は`variableNamesLikeThis `このように小文字で始まり、区切りは大文字を利用するスモールキャメルケースを利用しましょうか！ こっちの方がjavascriptLikeな書き方です！
![googleJavascriptStyleGuide](http://cou929.nu/data/google_javascript_style_guide/#id28)

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,41 @@
+			if (users.length !== 0) {
+				users.forEach(function(user){
+				appendUser(user);
+			});
```
特にif文のインデントは意識して変更するようにしてください

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,41 @@
+		user_list.append(html)
+	}
+
+	function appendNoUser(user) {
```
この関数って何をappendしているのでしょうか。。。
'NoUser'をappend?している、、というのがあまり直感的ではないような気がします
（これはカリキュラムが悪いのかな。。？）
→もし他に良さそうな名前が浮かんだら編集してみてください。。！

Hiro3nori3 replied:
"一致するユーザがいません"という文字列をビューに追加しています。他に良い名前が思いつきませんでした...

fumihumi replied:
なるほどですね！おっけいです！

fumihumi commented on app/views/groups/new.html.haml:
```diff
@@ -1,44 +1,29 @@
+				= f.text_field :name, class: "chat-group-form__input", placeholder: "グループ名を入力してください"
+		.chat-group-form__field.clearfix
+			.chat-group-form__field--left
+				%label.chat-group-form__label{:for => "chat_group_チャットメンバーを追加"} チャットメンバーを追加
```
ここは `f.label`にしましょう

fumihumi replied:
ここは`f.label`にしましょう！

fumihumi commented on app/views/groups/new.html.haml:
```diff
@@ -1,44 +1,29 @@
+				%label.chat-group-form__label{:for => "chat_group_チャットメンバーを追加"} チャットメンバーを追加
+			.chat-group-form__field--right
+				.chat-group-form__search.clearfix
+					%input#user-search-field.chat-group-form__input{:placeholder => "追加したいユーザー名を入力してください", :type => "text"}/
```
ここも`f.~~`にしましょう

fumihumi commented:
`group#edit`って編集してなくて大丈夫でしょうか？？？？？？？？？？

fumihumi commented:
descriptionの *#* の後はスペースあけないとマークダウン記法にならないですね！w

Commit:
search.jsの編集

・変数をキャメルケースで書くようにした
・ユーザー追加ボタンを押した際に、そのユーザーがビュー上に描画されるようにした
・if文のインデントを調節した

Commit:
new.html.hamlの編集

・エラーメッセージがあるときとないときでif文を追加した
・labelをf.labelを使って書き直した
・edit.html.hamlをnew.html.hamlの編集内容に合わせて作り直した

Commit:
edit.html.hamlの編集

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	function appendUserToGroup(user){
+		var user_id = $(user).data("user-id");
```
ここらへんも変数名意識してください！

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,64 @@
+
+
+	$("#user-search-field").on("keyup", function() {
+		var input = $("#user-search-field").val();
```
ここは引数の無名関数にイベントの引数"e"を渡してそこから値を取得する。
もしくはjsのthisなど使って入力値をもらいましょう！

jqueryでの"#user-search-field"へのdom探索は一回だけにしたいですね

Hiro3nori3 replied:
これ次のコミットまでお待ち下さいm(_ _)m

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,64 @@
+	});
+	$(document).on('click', '.user-search-add', function(){
+		var user = $(this).parent();
+		user.remove();
```
ここは変数に格納しないでそのままremoveで良さそうです。

今後、userっていうのを利用するなら変数に入れたいですが、特に使わないと思うので

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,64 @@
+	$(document).on("click",".user-search-remove",function(e){
+		e.preventDefault();
+		var user = $(this).parent();
+		user.remove();
```
こちらも同じですね

fumihumi reviewed (commented):
formforのなかでは、出来るだけビューヘルパー使いましょう！
inputタグなどです！

また、editとnewは同じフォームなので部分テンプレートにしましょう！

Hiro3nori3 commented:
質問なのですが、新しくグループを作成する際に、current_userのidの値をビューからコントローラーに渡すにはどうすればよいでしょうか？

Hiro3nori3 commented:
こちらなんですけど、新規グループ作成の際にビューチャットメンバーのところにログインユーザーを表示させるようにしたのですが、グループ編集の際に、その下に更にログインユーザーが表示されるようになってしまいました。編集の際に、group.users.each |user|の部分で、ログインユーザーが含まれないようにするにはどうすれば良いでしょうか？
[![Screenshot from Gyazo](https://gyazo.com/93946a6b600415ea1a4caa6bcce6d312/raw)](https://gyazo.com/93946a6b600415ea1a4caa6bcce6d312)
[![Screenshot from Gyazo](https://gyazo.com/47abb2edde0a582be52a963a517a85f8/raw)](https://gyazo.com/47abb2edde0a582be52a963a517a85f8)

Commit:
_group.scssの編集

・なぜかCSSがうまく効かないようになっていたため、カリキュラムからダウンロードしたものを貼り直した
・以前のものはひとまずコメントアウトした

Commit:
フォームの部分テンプレート化

・フォームを部分テンプレート化した
・hidden_fieldタグを用いて、データベースにデータが保存されるようにした
・新規グループ作成の画面でログインユーザーがチャットメンバーに表示されるようにした

Commit:
search.jsの修正

・変数名をJavascript-likeにした
・remove()の部分の記述をリファクタリングした

fumihumi commented on app/assets/javascripts/search.js:
```diff
@@ -0,0 +1,62 @@
+		var input = $("#user-search-field").val();
+		$.ajax({
+			type: 'GET',
+			url: '/users/search',
```
’/users/search’を変数に格納して変数として利用しましょうか！
そっちの方がわかりやすい気がします。

Hiro3nori3 replied:
'/users/search'を利用する場所がここしかないような気がするのですが、変数にする意味はあるのでしょうか...?

fumihumi commented on app/views/groups/_form.html.haml:
```diff
@@ -0,0 +1,39 @@
+			= f.label :text, class: "chat-group-form__label", value: "チャットメンバーを追加"
+		.chat-group-form__field--right
+			.chat-group-form__search.clearfix
+				%input#user-search-field.chat-group-form__input{placeholder: "追加したいユーザー名を入力してください", type: "text"}/
```
inputはf.text~などに変えましょう！

fumihumi commented:
> 編集の際に、group.users.each |user|の部分で、ログインユーザーが含まれないようにするにはどうすれば良いでしょうか？

これって `group.users`の特徴で、そのグループに属しているuserを取得していることが原因ですね。
なので、groupModelなどにスコープなどを用意してそれを、Viewから呼び出しましょう！

どんなデータが欲しいのかを今一度考えてSQLを作成しましょう！
SQLの`where`を編集することで取得できます！

Commit:
group.rbの編集

フォームでユーザーを表示させる際にログインユーザーを除くユーザーを表示させるようにした

Commit:
users_controllerの編集

・searchアクションの@usersをカレントユーザー以外を表示させるようにした

Commit:
修正 ver. 1

・inputタグをf.text_fieldに書き換えた
・search.jsのリファクタリング

Hiro3nori3 commented:
ページをリロードすると効かなかったJSファイルが効くようになるって原因としてどのようなことが考えられますかね...?
(追記）turbolinksの設定切ったら、いけました！
[![Screenshot from Gyazo](https://gyazo.com/2971ccc7aa93b80adad89a8a29c27e72/raw)](https://gyazo.com/2971ccc7aa93b80adad89a8a29c27e72)

fumihumi commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,9 @@
+class UsersController < ApplicationController
+	def search
+		@users = User.where('name LIKE(?)', "%#{params[:keyword]}%").where.not(id: current_user)
```
複雑なSQLを利用する場合はスコープを使ってください！

fumihumi commented on app/models/group.rb:
```diff
@@ -4,4 +4,8 @@ class Group < ApplicationRecord
 
 	validates :name, presence: true
+
+	def delete(current_user)
```
`delete`はメソッド名として不適切かなと感じます。
(https://ref.xaio.jp/ruby/classes/string/delete)
すでにString型にdeleteメソッドがあるので。

Commit:
turbolinks off

・turbo linksをoffにした

Commit:
スコープの利用

・複雑なSQL文をスコープでモデルに記述した

Commit:
omit(current_user)

・group.rbのメソッド名を変えた

Commit:
exclude(current_user)

fumihumi commented:
![](http://www.lgtm.in/p/CfmXykY4q)

Commit:
修正 ver.2

・細かな修正
